### PR TITLE
Rewrite "The Kubernetes network model"

### PR DESCRIPTION
**Pull request by danwinship**
Spun off from #41419 which in turn is spun off from #39890. Whee!

This _just_ focuses on improving the existing text, and doesn't (much) add anything to it (leaving that for the two older PRs).

The big problem with this page is that it is very very very old, and that shows in two main ways:

1. It is *super* focused on explaining Kubernetes networking to people who are used to 2015-era single-host docker networking, but it never actually explains that that's what it's contrasting against. I'm sure that when this was first written, it was incredibly helpful, but at this point, it is just confusing to the vast majority of readers.
2. Especially toward the end of the doc, the hyperlinks are very weird. It looks like this is the result of gradual refactoring of the docs; someone linked to document X for its discussion of topic Y, but then later the discussion of topic Y was moved into a different document, leaving the old link now pointing to the wrong docs, etc.

I tried to keep most of the existing information, but some of the "kubernetes-is-not-docker" type stuff just didn't really seem to fit well and got dropped. (But if anyone thinks I dropped something important, let me know and I'll try to find a place for it.)

The update completely reorganizes the existing information, essentially moving the old "Kubernetes networking addresses four concerns" list to the top and putting the "fundamental requirements on any networking implementation" list inside it. And I expanded heavily on the "requirements on any networking implementation" / "detail of the particular container runtime in use" comments to talk about what parts of the network model are core to k8s and what parts are implemented externally.

I'm sure there are style issues to be fixed... (It may need some better transitions in places, or subheadings?)

/cc @sftim @aojea @thockin

### CODE REVIEW

File: content/en/docs/concepts/services-networking/_index.md [hunk 1]
```diff
@@ -7,58 +7,90 @@ description: >
...
-  to be reachable from outside your cluster.
-  - [Ingress](/docs/concepts/services-networking/ingress/) provides extra functionality
-    specifically for exposing HTTP applications, websites and APIs.
-  - [Gateway API](/docs/concepts/services-networking/gateway/) is an {{<glossary_tooltip text="add-on" term_id="addons">}}
-    that provides an expressive, extensible, and role-oriented family of API kinds for modeling service networking.
-- You can also use Services to
-  [publish services only for consumption inside your cluster](/docs/concepts/services-networking/service-traffic-policy/).
+The Kubernetes network model is built out of several pieces:
+
+  * Each [`Pod`](/docs/concepts/workloads/pods/) in a cluster gets its
```
Comment (sftim) [inline content/en/docs/concepts/services-networking/_index.md:12]:
```suggestion
  * Each [Pod](/docs/concepts/workloads/pods/) in a cluster gets its
```

File: content/en/docs/concepts/services-networking/_index.md [hunk 2]
```diff
@@ -7,58 +7,90 @@ description: >
...
+    own unique cluster-wide IP address.
+
+      * A pod has its own private network namespace which is shared by
+        all of the containers within the pod. Processes running in
+        different containers in the same pod can communicate with each
+        other over `localhost`.
+
+  * The pod network (or "cluster network") handles communication
```
Comment (sftim) [inline content/en/docs/concepts/services-networking/_index.md:20]:
```suggestion
  * The _pod network_ (also called a cluster network) handles communication
```

File: content/en/docs/concepts/services-networking/_index.md [hunk 3]
```diff
@@ -7,58 +7,90 @@ description: >
...
+    between pods:
+
+      * All pods can communicate with all other pods, whether on the
+        same [node](/docs/concepts/architecture/nodes/) or on
+        different nodes, without NAT.
+
+          * (For platforms that support host-network pods (e.g.
+            Linux), this rule applies to host-network pods as well.)
```
Comment (sftim) [inline content/en/docs/concepts/services-networking/_index.md:28]: Aligning to the [style guide](https://kubernetes.io/docs/contribute/style/style-guide/):
```suggestion
      * All pods have a route available to all other pods, whether two Pods are on the
        same [node](/docs/concepts/architecture/nodes/) or on
        different nodes. This is a direct means to communicate (no NAT).

          * (For platforms that support host-network pods (Linux),
          this rule applies to host-network pods as well.)
```
Pods may or may not be able to actually communicate, due to:
- packet filtering (and similar controls)
- misconfiguration, I suppose
- outages

File: content/en/docs/concepts/services-networking/_index.md [hunk 4]
```diff
@@ -7,58 +7,90 @@ description: >
...
+
+      * Agents on a node (e.g. system daemons, kubelet) can communicate
+        with all pods on that node.
```
Comment (sftim) [inline content/en/docs/concepts/services-networking/_index.md:31]:
```suggestion
      * Agents on a node (such as system daemons, kubelet) can communicate
        with all pods on that node.
```

File: content/en/docs/concepts/services-networking/_index.md [hunk 5]
```diff
@@ -7,58 +7,90 @@ description: >
...
+
+  * The [`Service`](/docs/concepts/services-networking/service/) API
```
Comment (sftim) [inline content/en/docs/concepts/services-networking/_index.md:33]:
```suggestion
  * The [Service](/docs/concepts/services-networking/service/) API
```

File: content/en/docs/concepts/services-networking/_index.md [hunk 6]
```diff
@@ -7,58 +7,90 @@ description: >
...
+    lets you provide a stable "cluster IP" for a service implemented
```
Comment (sftim) [inline content/en/docs/concepts/services-networking/_index.md:34]:
```suggestion
    lets you provide a stable (long lived) IP address or hostname for a service implemented
```

File: content/en/docs/concepts/services-networking/_index.md [hunk 7]
```diff
@@ -7,58 +7,90 @@ description: >
...
+    by one or more backend pods, where the individual pods making up
+    the service can change over time.
+
+      * Kubernetes creates
+        [`EndpointSlice`](content/en/docs/concepts/services-networking/endpoint-slices/)
```
Comment (sftim) [inline content/en/docs/concepts/services-networking/_index.md:39]:
```suggestion
        [EndpointSlice](/docs/concepts/services-networking/endpoint-slices/)
```

File: content/en/docs/concepts/services-networking/_index.md [hunk 8]
```diff
@@ -7,58 +7,90 @@ description: >
...
+
+      * Agents on a node (e.g. system daemons, kubelet) can communicate
+        with all pods on that node.
+
+  * The [`Service`](/docs/concepts/services-networking/service/) API
+    lets you provide a stable "cluster IP" for a service implemented
+    by one or more backend pods, where the individual pods making up
+    the service can change over time.
+
+      * Kubernetes creates
```
Comment (sftim) [inline content/en/docs/concepts/services-networking/_index.md:38]:
```suggestion
      * Kubernetes automatically manages
```

File: content/en/docs/concepts/services-networking/_index.md [hunk 9]
```diff
@@ -7,58 +7,90 @@ description: >
...
+        backing a Service.
+
+      * A service proxy implementation, takes care of converting the
+        `Service` and `EndpointSlice` objects into packet-rewriting
+        rules using appropriate operating system APIs.
+
+  * The [`Gateway`](/docs/concepts/services-networking/gateway/) API
+    (or its predecessor,
+    [`Ingress`](/docs/concepts/services-networking/ingress/)) allows
+    you to make Services accessible from outside the cluster.
```
Comment (sftim) [inline content/en/docs/concepts/services-networking/_index.md:50]:
```suggestion
  * The [Gateway](/docs/concepts/services-networking/gateway/) API
    (or its predecessor,
    [Ingress](/docs/concepts/services-networking/ingress/)) allows
    you to make Services accessible to clients that are outside the cluster.
```

File: content/en/docs/concepts/services-networking/_index.md [hunk 10]
```diff
@@ -7,58 +7,90 @@ description: >
...
+
+      * A simpler, but less-configurable, mechanism for cluster
+        ingress is available via the Service API's [`type:
+        LoadBalancer`](/docs/concepts/services-networking/service/#loadbalancer),
+        when using a supported {{< glossary_tooltip term_id="cloud-provider">}}.
+
+  * [`NetworkPolicy`](/docs/concepts/services-networking/network-policies)
```
Comment (sftim) [inline content/en/docs/concepts/services-networking/_index.md:57]:
```suggestion
  * [NetworkPolicy](/docs/concepts/services-networking/network-policies) is a built-in
    Kubernetes API that
```

File: content/en/docs/concepts/services-networking/_index.md [hunk 11]
```diff
@@ -7,58 +7,90 @@ description: >
...
+    allows you to control traffic between pods, or between pods and
+    the outside world.
```
Comment (sftim) [inline content/en/docs/concepts/services-networking/_index.md:59]:
```suggestion
    the outside world. Your cluster **only** enforces NetworkPolicy if the cluster network supports
    that kind of enforcement; read the NetworkPolicy documentation to learn more about this
    restriction.
```

File: content/en/docs/concepts/services-networking/_index.md [hunk 12]
```diff
@@ -7,58 +7,90 @@ description: >
...
+
+In older container systems, there was no automatic connectivity
+between containers on different hosts, and so it was often necessary
+to explicitly create links between containers, or to map container
+ports to host ports to make them reachable by containers on other
+hosts. This is not needed in Kubernetes; Kubernetes's model is that
+`Pods` can be treated much like VMs or physical hosts from the
```
Comment (sftim) [inline content/en/docs/concepts/services-networking/_index.md:66]:
```suggestion
pods can be treated much like VMs or physical hosts from the
```

File: content/en/docs/concepts/services-networking/_index.md [hunk 13]
```diff
@@ -7,58 +7,90 @@ description: >
...
+  * Pod network namespace setup is handled by system-level software
+    implementing the [Container Runtime
+    Interface](/docs/concepts/architecture/cri.md).
+
+  * The pod network itself is managed by a [pod network
+    implementation](/docs/concepts/cluster-administration/addons/#networking-and-network-policy).
+    On Linux, most container runtimes use the
+    {{< glossary_tooltip text="Container Networking Interface (CNI)" term_id="cni" >}}
+    to interact with the pod network implementation, so these
+    implementations are often called "CNI plugins"
```
Comment (sftim) [inline content/en/docs/concepts/services-networking/_index.md:83]:
```suggestion
    implementations are often called _CNI plugins_.
```

File: content/en/docs/concepts/services-networking/_index.md [hunk 14]
```diff
@@ -7,58 +7,90 @@ description: >
...
+
+  * Kubernetes provides a default implementation of service proxying,
+    called {{< glossary_tooltip term_id="kube-proxy">}}, but some pod
+    network implementations instead use their own service proxy that
+    is more tightly integrated with the rest of the implementation.
+
+  * There are many [implementations of the Gateway
+    API](https://gateway-api.sigs.k8s.io/implementations/), some of
+    which are specific to particular cloud environments, some more
+    focused on "bare metal" environments, and others more generic.
```
Comment (sftim) [inline content/en/docs/concepts/services-networking/_index.md:94]: Please add a (repeated) disclaimer / admonition about NetworkPolicy and note that
it may not be enforced even though you can define one.

(aside: wouldn't it be great if there was a default warning, that gets turned off when a network plugin marks NetworkPolicy - the API - as enforced?)

Also:

```suggestion

## {{% heading "whatsnext" %}}

```

**Review by sftim — COMMENTED** [review]
Thanks. We've updated the style guide since this page was last updated; please follow the current style guide around how we write API kinds (and also, ideally, avoid Latin abbreviations).

/lgtm

This could merge as-is; it's an obvious improvement on what we have now.

File: content/en/docs/concepts/services-networking/_index.md [hunk 15]
```diff
@@ -30,31 +30,32 @@ The Kubernetes network model is built out of several pieces:
       * Agents on a node (e.g. system daemons, kubelet) can communicate
         with all pods on that node.
 
-  * The [`Service`](/docs/concepts/services-networking/service/) API
-    lets you provide a stable "cluster IP" for a service implemented
+  * The [Service](/docs/concepts/services-networking/service/) API
+    lets you provide a stable (long lived) IP address or hostname for a service implemented
```
Comment (aojea) [inline content/en/docs/concepts/services-networking/_index.md:34]: what is the "long lived" meaning here? it can be short lived, is about is `reserved` ? or something like that?

File: content/en/docs/concepts/services-networking/_index.md [hunk 16]
```diff
@@ -80,7 +81,7 @@ corresponding functionality is provided by external components:
     On Linux, most container runtimes use the
     {{< glossary_tooltip text="Container Networking Interface (CNI)" term_id="cni" >}}
     to interact with the pod network implementation, so these
-    implementations are often called "CNI plugins"
+    implementations are often called _CNI plugins_.
```
Comment (aojea) [inline content/en/docs/concepts/services-networking/_index.md:84]: I'm surprised you didn't added "network plugins"  here too :)

**Comment by aojea** [comment]
LGTM module one comment about the meaning of "long lived" in services, but because I don't understand well the meaning of that expression there

File: content/en/docs/concepts/services-networking/_index.md (hunk 15 above)
Comment (sftim) [reply content/en/docs/concepts/services-networking/_index.md:34]: It can outlive one Pod, is what I want to convey.

**Comment by sftim** [comment]
I judge that even where there's room to improve further, this makes the docs better than they are as they stand.

/lgtm

**Comment by danwinship** [comment]
/hold
(still planning to address the remaining comments, I just didn't get to that yet)

**Comment by salaxander** [comment]
Thanks for all the work on this @danwinship!

**Comment by salaxander** [comment]
For next week's PR wrangler - I think this looks great so far, when Dan is happy with the state of it, I think it's good to merge

File: content/en/docs/concepts/services-networking/_index.md (hunk 3 above)
Comment (danwinship) [reply content/en/docs/concepts/services-networking/_index.md:28]: I brought back the "barring intentional network segmentation" text from before. That applies to the "agents on a node" subclause too, so I put it at the "the pod network handles communication" level rather than the "all pods can communicate" level.

On the Linux thing, it felt weird to be calling out the Linux behavior, which is the expected behavior, rather than the Windows behavior, which is considered wrong, so I flipped that.

File: content/en/docs/concepts/services-networking/_index.md (hunk 11 above)
Comment (danwinship) [reply content/en/docs/concepts/services-networking/_index.md:60]: (I didn't add anything here, but I did add it below.)

File: content/en/docs/concepts/services-networking/_index.md (hunk 14 above)
Comment (danwinship) [reply content/en/docs/concepts/services-networking/_index.md:95]: > (aside: wouldn't it be great if there was a default warning, that gets turned off when a network plugin marks NetworkPolicy - the API - as enforced?)

ah, but then all existing plugins would erroneously trigger the warning until they got updated :slightly_frowning_face:. But we learned from our mistakes at least; AdminNetworkPolicy is a CRD that is installed by the ANP implementation, so if the API is present, then it's implemented. (And the plan, hopefully, is to eventually add "DeveloperNetworkPolicy" as a replacement for NetworkPolicy, which would work the same way, and then we can deprecate NetworkPolicy, and then we'll all have jetpacks and flying cars.)

Until then, the existence of https://github.com/kubernetes-sigs/kube-network-policies means there's no good reason to not implement NetworkPolicy these days so hopefully lack-of-NP support will become less common...

File: content/en/docs/concepts/services-networking/_index.md [hunk 17]
```diff
@@ -7,58 +7,102 @@ description: >
...
+    the service can change over time.
+
+      * Kubernetes automatically manages
+        [EndpointSlice](/docs/concepts/services-networking/endpoint-slices/)
+        objects to provide information about the pods currently
+        backing a Service.
+
+      * A service proxy implementation, takes care of converting the
+        `Service` and `EndpointSlice` objects into packet-rewriting
+        rules using appropriate operating system APIs.
```
Comment (aojea) [inline content/en/docs/concepts/services-networking/_index.md:47]: There can be service proxy implemenations that are userspace can we rephrase this to be less low level "A service proxy implementation uses the Services and EndpointSlices object to program the dataplane using appropriate operating system APIs" or something like that

File: content/en/docs/concepts/services-networking/_index.md [hunk 18]
```diff
@@ -7,58 +7,102 @@ description: >
...
+pods can be treated much like VMs or physical hosts from the
+perspectives of port allocation, naming, service discovery, load
+balancing, application configuration, and migration.
+
+Only a few parts of this model are implemented by Kubernetes itself.
+For the other parts, Kubernetes defines the APIs, but the
+corresponding functionality is provided by external components, some
+of which are optional:
+
+  * Pod network namespace setup is handled by system-level software
```
Comment (aojea) [inline content/en/docs/concepts/services-networking/_index.md:78]:
```suggestion
  * Pod network namespace setup is handled by container runtimes that use the [Container Runtime
    Interface] to communicate with the kubelet
```

File: content/en/docs/concepts/services-networking/_index.md (hunk 14 above)
Comment (sftim) [reply content/en/docs/concepts/services-networking/_index.md:104]: We still have to let people know about the risk, so long as there is a network plugin that enables packet networking but not NetworkPolicy.

File: content/en/docs/concepts/services-networking/_index.md [hunk 19]
```diff
@@ -7,58 +7,102 @@ description: >
...
+    by one or more backend pods, where the individual pods making up
+    the service can change over time.
+
+      * Kubernetes automatically manages
+        [EndpointSlice](/docs/concepts/services-networking/endpoint-slices/)
+        objects to provide information about the pods currently
+        backing a Service.
+
+      * A service proxy implementation, takes care of converting the
+        `Service` and `EndpointSlice` objects into packet-rewriting
```
Comment (sftim) [inline content/en/docs/concepts/services-networking/_index.md:46]:
```suggestion
        Service and EndpointSlice objects into packet-rewriting
```

File: content/en/docs/concepts/services-networking/_index.md [hunk 20]
```diff
@@ -7,58 +7,102 @@ description: >
...
+    lets you provide a stable (long lived) IP address or hostname for a service implemented
+    by one or more backend pods, where the individual pods making up
+    the service can change over time.
+
+      * Kubernetes automatically manages
+        [EndpointSlice](/docs/concepts/services-networking/endpoint-slices/)
+        objects to provide information about the pods currently
+        backing a Service.
+
+      * A service proxy implementation, takes care of converting the
```
Comment (sftim) [inline content/en/docs/concepts/services-networking/_index.md:45]:
```suggestion
      * A _service proxy_ implementation takes care of converting information from
```

File: content/en/docs/concepts/services-networking/_index.md [hunk 21]
```diff
@@ -7,58 +7,102 @@ description: >
...
+balancing, application configuration, and migration.
+
+Only a few parts of this model are implemented by Kubernetes itself.
+For the other parts, Kubernetes defines the APIs, but the
+corresponding functionality is provided by external components, some
+of which are optional:
+
+  * Pod network namespace setup is handled by system-level software
+    implementing the [Container Runtime
+    Interface](/docs/concepts/architecture/cri.md).
```
Comment (sftim) [inline content/en/docs/concepts/services-networking/_index.md:80]:
```suggestion
implementing the
{{< glossary_tooltip text="Container Runtime Interface" term_id="cri" >}}.
```
(also fixes a broken link)

File: content/en/docs/concepts/services-networking/_index.md [hunk 22]
```diff
@@ -7,58 +7,102 @@ description: >
...
+    (or its predecessor,
+    [Ingress](/docs/concepts/services-networking/ingress/)) allows
+    you to make Services accessible to clients that are outside the cluster.
+
+      * A simpler, but less-configurable, mechanism for cluster
+        ingress is available via the Service API's [`type:
+        LoadBalancer`](/docs/concepts/services-networking/service/#loadbalancer),
+        when using a supported {{< glossary_tooltip term_id="cloud-provider">}}.
+
+  * [NetworkPolicy](/docs/concepts/services-networking/network-policies) is a built-in
```
Comment (sftim) [inline content/en/docs/concepts/services-networking/_index.md:59]: (nit)
```suggestion
  * [NetworkPolicy](/docs/concepts/services-networking/network-policies/) is a built-in
```

File: content/en/docs/concepts/services-networking/_index.md [hunk 23]
```diff
@@ -7,58 +7,102 @@ description: >
...
-  to be reachable from outside your cluster.
-  - [Ingress](/docs/concepts/services-networking/ingress/) provides extra functionality
-    specifically for exposing HTTP applications, websites and APIs.
-  - [Gateway API](/docs/concepts/services-networking/gateway/) is an {{<glossary_tooltip text="add-on" term_id="addons">}}
-    that provides an expressive, extensible, and role-oriented family of API kinds for modeling service networking.
-- You can also use Services to
-  [publish services only for consumption inside your cluster](/docs/concepts/services-networking/service-traffic-policy/).
+The Kubernetes network model is built out of several pieces:
+
+  * Each [Pod](/docs/concepts/workloads/pods/) in a cluster gets its
```
Comment (sftim) [inline content/en/docs/concepts/services-networking/_index.md:12]: (extra credit / nit)
```suggestion
  * Each {{< glossary_tooltip text="pod" term_id="pod" >}} in a cluster gets its
```

**Review by sftim — COMMENTED** [review]
If we merge this as-is, the docs become better.

/lgtm

File: content/en/docs/concepts/services-networking/_index.md (hunk 17 above)
Comment (danwinship) [reply content/en/docs/concepts/services-networking/_index.md]: This is just an overview, and given that all official proxy implementations and commonly-used third-party proxy implementations rewrite packets, it seems reasonable to say it this way.

Comment (sftim) [reply content/en/docs/concepts/services-networking/_index.md]: I agree with @danwinship, and also the data plane programming might be done by something even more custom (for example: cloud provider APIs; compiling P4 code and reconfiguring external hardware).

**Comment by sftim** [comment]
(still)
/lgtm

File: content/en/docs/concepts/services-networking/_index.md (hunk 17 above)
Comment (aojea) [reply content/en/docs/concepts/services-networking/_index.md]: fair enough

File: content/en/docs/concepts/services-networking/_index.md [hunk 24]
```diff
@@ -7,58 +7,102 @@ description: >
...
+    by one or more backend pods, where the individual pods making up
+    the service can change over time.
+
+      * Kubernetes automatically manages
+        [EndpointSlice](/docs/concepts/services-networking/endpoint-slices/)
+        objects to provide information about the pods currently
+        backing a Service.
+
+      * A _service proxy_ implementation takes care of converting information from
+        Service and EndpointSlice objects into packet-rewriting
```
Comment (thockin) [inline content/en/docs/concepts/services-networking/_index.md:46]: s/rewriting/routing ?

File: content/en/docs/concepts/services-networking/_index.md [hunk 25]
```diff
@@ -7,58 +7,102 @@ description: >
...
+        rules using appropriate operating system APIs.
+
+  * The [Gateway](/docs/concepts/services-networking/gateway/) API
+    (or its predecessor,
+    [Ingress](/docs/concepts/services-networking/ingress/)) allows
+    you to make Services accessible to clients that are outside the cluster.
+
+      * A simpler, but less-configurable, mechanism for cluster
```
Comment (thockin) [inline content/en/docs/concepts/services-networking/_index.md:54]: Are we comfortable implying Service LB is obsolete?  Does L4 Gateway handle EVERYTHING that LB does?  I don't think so...

**Review by thockin — COMMENTED** [review]
Thanks, this was WAY overdue

File: content/en/docs/concepts/services-networking/_index.md (hunk 25 above)
Comment (danwinship) [reply content/en/docs/concepts/services-networking/_index.md:54]: What change would you recommend?

I definitely wasn't trying to imply that it's obsolete (and note that the old version of this page doesn't link to the Service LB docs at all, so this is actually making that feature _more_ prominent).

But if we're going to have a single top-level answer for "how cluster ingress works in Kubernetes", it should be Gateway, right?

(I could demote the Ingress reference to a nested bullet point... or remove it.)

File: content/en/docs/concepts/services-networking/_index.md (hunk 24 above)
Comment (danwinship) [reply content/en/docs/concepts/services-networking/_index.md:46]: OK, Antonio wanted to rewrite this point too....

File: content/en/docs/concepts/services-networking/_index.md [hunk 26]
```diff
@@ -7,58 +7,103 @@ description: >
...
+
+      * Kubernetes automatically manages
+        [EndpointSlice](/docs/concepts/services-networking/endpoint-slices/)
+        objects to provide information about the pods currently
+        backing a Service.
+
+      * A service proxy implementation monitors the `Service` and
+        `EndpointSlice` objects, and programs the data plane to route
+        service traffic to its backends, by using operating system or
+        cloud provider APIs to intercept or rewrite packets.
```
[the remark this answers is not in the document]
Comment (thockin) [reply content/en/docs/concepts/services-networking/_index.md:48]: LGTM

File: content/en/docs/concepts/services-networking/_index.md (hunk 25 above)
Comment (thockin) [reply content/en/docs/concepts/services-networking/_index.md:55]: >  if we're going to have a single top-level answer for "how cluster ingress works in Kubernetes", it should be Gateway, right?

I can't see how that is true until L4 gateway is as complete as service LB?  @robscott 

As for wording here, something like:

"An alternative mechanism for exposing a single Service is the Service API's type: LoadBalancer.  It is simpler than Gateway, but less powerful."

Or something?  I know we don't want to make this "at the time of this writing" but the reality is that, at the time of this writing Gateway does not fully replace type: LB.

Comment (sftim) [reply content/en/docs/concepts/services-networking/_index.md:55]: I think this PR makes a better compromise than the page as it stands (and obviously so), so I strongly support merging as-is and then iterating.

I don't see anything that is so wrong that it would seriously mislead readers or get in the way of their learning.

**Comment by reylejano** [comment]
/label refactor
/sig network
This PR is a wonderful improvement. The Kubernetes network model is easier to understand with this PR (imo). 

This PR has reviews from SIG Network co-chair and tech lead.
/approve

**Comment by aojea** [comment]
LGTM, @thockin for the final approval

**Comment by sftim** [comment]
I can't imagine @thockin objecting at this point, so:

/lgtm

**Comment by thockin** [comment]
LGTM
